teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,212 @@
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Option 1: Allowing Users to Specify Redirection Paths](#option-1-allowing-users-to-specify-redirection-paths)
+    - [Example Usage](#example-usage)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Option 2: Redirecting to Canonical Conventional Paths](#option-2-redirecting-to-canonical-conventional-paths)
+    - [Example Usage](#example-usage-1)
+    - [Risks and Mitigations](#risks-and-mitigations-1)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit output stream redirection. This is not even possible if the image does not come with a shell.
+
+To achieve the functionality of output stream redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to redirect stdout/stderr to.
+
+## Motivation
+
+This TEP extends Tekton Pipelines to support the following use cases:
+
+* Allow Task authors to run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([tektoncd/pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)). Generally speaking, this allows Task authors to apply [Unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy) to container images using steps and make multiple images work together in a Task.
+
+* Allow Task authors to run images not controlled by Task authors and still be able to use Task results (and potentially other path-based features such as output resources). It is common for users to use some third-party “official” utility images instead of maintaining their own fork to include a shell, and thus impossible to ensure that there is always a shell to use the script for that shell. In the current Tekton API, Task results cannot be used with certain images (e.g., images without a shell and whose entrypoint does not provide an option to write outputs to files), so there is an incompleteness in the API, and this TEP proposes a way to address this limitation.
+
+* Allow tool developers to create CI pipeline tooling that can use Task results on *any* image specified by the end users without the above limitations. As Task authors, one can choose which images to use to work around the limitations. But as a tool developer, one cannot or should not control what images their end users can use, and cannot assume any details about the images.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step.
```

Comment:
i have a question around supporting stderr - i dont want to push back strongly against it cuz if we're doing stdout, it's not a lot of work to do stderr, but i do wonder if we actually need it and if there is a (reasonable?) possibility to start with just stdout

From what I can tell, none of the use cases above require stderr, and if we look at how Unix processes are designed, it looks like [passing stderr between them is not the norm either](https://en.wikipedia.org/wiki/Pipeline_(Unix)#Error_stream)

There are some other uses cases that WOULD require capturing stderr tho, maybe it would make sense to include them here too e.g. @coryrc and prow folks want a way to capture the logs from running containers and upload them.

and/or we could do just stdout and deal with stderr differently

Reply:
im also wondering if stderr is more of a runtime concern - i.e. who wants to express that stderr should be captured and dealt with in some particular way? im thinking about if you put a Task in the catalog, what are the chances that you'll want a step in that Task to be dealing with stderr - or is it more likely that at runtime when you use that Task, you'll want a way to grab logs from that Task?

Reply:
> Here it is similar, we (tekton contributors, developers) don't know how stderr will be use by our users (and the binaries) ; we can't make too much assumption on how the author of the thing running in the step will use stdout and stderr to pass informations.

To me it comes down to the use cases we're trying to address - i think we're specifically trying to address use cases where folks are trying to use tools that were written to expect to be able to pipe stdout into other processes.

Capturing stderr seems like a totally different set of use cases to me.

Reply:
I have no problem leaving stderr out of the scope of this TEP. I added it simply to address your initial feedback on slack before I started this TEP ;) I agree that in terms of use cases, capturing stderr is probably less useful.

There is a potential use case though: error handling with command-line results. Since most commands exit `EXIT_FAILURE` if there is any application-level failure, it's common (although IMO not advisable) to process the stderr contents through e.g. `grep` to further identify the exact errors.

Reply:
> To me it comes down to the use cases we're trying to address - i think we're specifically trying to address use cases where folks are trying to use tools that were written to expect to be able to pipe stdout into other processes.

I don't see that use case only. For me, we are trying to address use cases where **folks want to write somewhere the output of the task, for re-use or storage, without requiring to have to tooling inside the image** (aka rebuild an existing image specifically to run in tekton). That's the only *assumption* I make on the user for this feature. The fact that it covers the "tools that were written to expect to be able to pipe stdout into other processes" is a good thing, but it's not the sole use case we should cover.

nit : most of tools that are written to output something, are able to use piping and all. Very few tools are written *for* being used in pipes — `tr`, `cut`, … are, but `grep`, `git`, are not built for it, they benefit from it.